File: framework/engine/systemml/link.cpp
/*
________________________________________________________________

	This file is part of BRAHMS
	Copyright (C) 2007 Ben Mitchinson
	URL: http://brahms.sourceforge.net

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
________________________________________________________________

	Subversion Repository Information (automatically updated on commit)

	$Id:: link.cpp 2256 2009-10-31 15:45:54Z benjmitch         $
	$Rev:: 2256                                                $
	$Author:: benjmitch                                        $
	$Date:: 2009-10-31 15:45:54 +0000 (Sat, 31 Oct 2009)       $
________________________________________________________________

*/



#include "systemml.h"


namespace brahms
{
	namespace systemml
	{


		Link::Link(string p_name, string p_path, string p_src, string p_dst, string p_lag)
		{
			name = p_name;

			setSrc(p_path + p_src);
			setDst(p_path + p_dst);

			DOUBLE n = brahms::text::s2n(p_lag.c_str());
			lag = n;
			if (n == brahms::text::S2N_FAILED || n != lag)
				ferr << E_SYSTEM_FILE << "invalid lag in Link";

			//	make inferences, as necessary - currently we don't support set
			//	transport, so we only support the following:
			//
			//	port to port					A>P to B<Q
			//									A>>X>P to B<<Y<Q
			if (src.type == ST_OUTPUT_PORT && dst.type == ST_INPUT_PORT)
			{
				//	ok
				dstPortNameInferred = false;
			}

			//	port to set						A>P to B<<Y (to <<Y<P)
			//									A>>X>P to B<<Y (to <<Y<P)
			else if (src.type == ST_OUTPUT_PORT && dst.type == ST_INPUT_SET)
			{
				//	infer portname and change dst type
				dst.portName = src.portName;
				dst.type = ST_INPUT_PORT;
				dstPortNameInferred = true;
			}

			//	port to process					A>P to B (to <P)
			//									A>>X>P to B (to <P)
			else if (src.type == ST_OUTPUT_PORT && dst.type == ST_PROCESS)
			{
				//	infer portname and change dst type
				dst.portName = src.portName;
				dst.type = ST_INPUT_PORT;
				dstPortNameInferred = true;
			}

			//	otherwise, raise a hand-wavy error
			else
			{
				ferr << E_SYSTEMML << "Link protocol either invalid or not currently supported in \"" << string(*this) << "\"";
			}
		}

		void Link::setSrc(string p_src)
		{
			src.parse(p_src.c_str(), ST_NULL, "Src");

			//	check src is set to the only cases we currently handle
			if (src.type != ST_OUTPUT_PORT)
				ferr << E_SYSTEM_FILE << "Link Src must be an output port (currently, set links are not supported) (was \"" << p_src << "\")";
		}

		void Link::setDst(string p_dst)
		{
			dst.parse(p_dst.c_str(), ST_NULL, "Dst");

			//	check dst is set to the only cases we currently handle
			if (dst.type != ST_INPUT_PORT && dst.type != ST_INPUT_SET && dst.type != ST_PROCESS)
				ferr << E_SYSTEM_FILE << "Link Dst must be an input port, set, or process (was \"" << p_dst << "\")";
		}

		string Link::getSrc()
		{
			return string(src);
		}

		string Link::getDst()
		{
			return string(dst);
		}

		UINT32 Link::getLag()
		{
			return lag;
		}

		string Link::getSrcProcessName()
		{
			return src.processName;
		}

		string Link::getDstProcessName()
		{
			return dst.processName;
		}

		Link::operator string() const
		{
			return string(src) + " ==> " + string(dst);
		}


	}
}

File: framework/bindings/995/@sml_system/link.m
%
% sys = sys.link(src, dst[, lag])
% sys = sys.link(name, src, dst[, lag])
%
%   Add a link to the system from src to dst, where src and
%   dst are SystemML identifiers. Optionally, provide a name
%   for the link itself as a reference (often, naming a link
%   will serve no purpose). The optional extra parameter lag
%   specifies the lag on the link, which defaults to unity.
%
%
% __________________________________________________________
% This function is part of the "SystemML Toolbox".


%__________________________________________________________________________
%
% This file is part of the SystemML Toolbox
% Copyright (C) 2007 Ben Mitch(inson)
% URL: http://sourceforge.net/projects/systemml
%
% This program is free software; you can redistribute it and/or modify it
% under the terms of the GNU General Public License as published by the
% Free Software Foundation; either version 2 of the License, or (at your
% option) any later version.
%
% This program is distributed in the hope that it will be useful, but
% WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
% General Public License for more details.
%
% You should have received a copy of the GNU General Public License along
% with this program; if not, write to the Free Software Foundation, Inc.,
% 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
%__________________________________________________________________________
%
% $Id:: link.m 2397 2009-11-18 21:20:07Z benjmitch                       $
% $Rev:: 2397                                                            $
% $Author:: benjmitch                                                    $
% $Date:: 2009-11-18 21:20:07 +0000 (Wed, 18 Nov 2009)                   $
%__________________________________________________________________________
%

function sys = link(sys, varargin)

% link(sys, src, dst[, lag])
% link(sys, name, src, dst[, lag])

% number of text arguments coming before a numeric or end of args
% indicates which form we're using
if nargin < 3
	error('not enough arguments');
end
if nargin == 3
	% empty name has special meaning: unnamed
	name = '';
	src = varargin{1};
	dst = varargin{2};
	lag = 1;
elseif nargin == 4
	if isnumeric(varargin{3})
		name = '';
		src = varargin{1};
		dst = varargin{2};
		lag = varargin{3};
	else
		name = varargin{1};
		src = varargin{2};
		dst = varargin{3};
		lag = 1;
	end
elseif nargin == 5
		name = varargin{1};
		% empty name has special meaning: unnamed
		src = varargin{2};
		dst = varargin{3};
		lag = varargin{4};
else
	error('invalid usage');
end

if ~ischar(name) | ~ischar(src) | ~ischar(dst) | ~isnumeric(lag)
	error('invalid usage');
end

validatename(sys, name, true);

link = [];
link.name = name;
link.src = src;
link.dst = dst;
link.lag = lag;
N = length(sys.link);
sys.link{N+1} = link;


% empty string does not place a constraint - can be used as
% name for multiple expose/link objects
if ~isempty(name)
	sys.namesinuse.(name) = 'link';
end



